src/main/java/com/griddynamics/cd/nrp/internal/uploading/impl/ArtifactUpdateApiClientImpl.java

```diff
@@ -50,23 +54,83 @@
                 configurationsManager.getConfiguration().getRequestsSendingThreadsCount(),
                 configurationsManager.getConfiguration().getRequestsSendingThreadsCount(),
-                30, TimeUnit.SECONDS, queue);
+                30,
+                TimeUnit.SECONDS,
+                new LinkedBlockingQueue<Runnable>(
+                        configurationsManager.getConfiguration().getRequestsQueueSize())
+        );
+        initBackgroundWorkers(configurationsManager);
+    }
+
+    private void initBackgroundWorkers(ConfigurationsManager configurationsManager) {
+        int requestsSendingThreadsCount = configurationsManager.getConfiguration()
+                .getRequestsSendingThreadsCount();
+        ExecutorService executorService = Executors.newFixedThreadPool(requestsSendingThreadsCount);
+        for (int i = 0; i < requestsSendingThreadsCount; i++) {
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    while (!Thread.currentThread().isInterrupted()) {
+                        try {
+                            final ArtifactMetaInfo artifactMetaInfo = fileBlockingQueue.peek();
+                            sendRequest(artifactMetaInfo);
+                            fileBlockingQueue.take();
+                        } catch (Exception e) {
+                            log.error(e.getMessage(), e);
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    private FileBlockingQueue initFileBlockingQueue(ConfigurationsManager configurationsManager) {
+        BlockingQueue<ArtifactMetaInfo> blockingQueue =
+                new LinkedBlockingQueue<>(configurationsManager.
+                        getConfiguration().getRequestsQueueSize());
+        String blockingQueueDumpFileName = configurationsManager.getConfiguration().getQueueDumpFileName();
+        FileBlockingQueue retVal = new FileBlockingQueue(blockingQueue,
+                blockingQueueDumpFileName);
+        try {
+            File blockingQueueInputFile = new File(blockingQueueDumpFileName);
+            if (blockingQueueInputFile.exists()) {
+                JAXBContext jaxbContext = JAXBContext.newInstance(ArtifactMetaInfoQueueDump.class);
+                Unmarshaller unmarshaller = jaxbContext.createUnmarshaller();
+                ArtifactMetaInfoQueueDump unmarshal = (ArtifactMetaInfoQueueDump) unmarshaller.unmarshal(blockingQueueInputFile);
+                for (ArtifactMetaInfo artifactMetaInfo : unmarshal.getArtifactMetaInfos()) {
+                    offerRequest(artifactMetaInfo);
+                }
+            }
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+        return retVal;
+    }
+
+    @Override
+    public void offerRequest(ArtifactMetaInfo artifactMetaInfo) {
+        try {
+            fileBlockingQueue.offer(artifactMetaInfo, 30, TimeUnit.SECONDS);
```

Comment:
Could you please extract 30 into a constant and add JavaDocs explaining why it's set to 30?